src/backend/pipeline/cont_adhoc_mgr.c

```diff
@@ -0,0 +1,205 @@
+#include "postgres.h"
+
+#include "pipeline/cont_adhoc_mgr.h"
+#include "storage/shmem.h"
+#include "storage/lwlock.h"
+#include "storage/proc.h"
+#include "miscadmin.h"
+#include "pipeline/tuplebuf.h"
+#include "access/xact.h"
+#include "catalog/pipeline_stream_fn.h"
+#include "catalog/pipeline_query_fn.h"
+
+/*
+ * The Adhoc Manager takes care of managing a small number of adhoc backends.
+ *
+ * It is responsible for initializing and configuring the shared memory for 
+ * backend -> backend comms wrt adhoc queries.
+ *
+ * It contains some functions necessary for synchronizing adhoc inserts and 
+ * removals from the AdhocTupleBuffer.
+ */
+
+/*
+ * The underlying data structure is one array of size 
+ * (max_worker_processes).
+ *
+ * Holes are allowed to form in the array. An empty slot is marked by
+ * group.db_oid being zero.
+ *
+ * Assumption is that max number of adhocs is small (32), and scanning it is
+ * fast.
+ */
+typedef struct AdhocShmemStruct
+{
+	int counter; /* always incrementing */
+	ContQueryProcGroup groups[1]; /* dynamically allocated */
+} AdhocShmemStruct;
+
+static AdhocShmemStruct *AdhocShmem;
+
+static size_t
+adhoc_size_required()
+{
+	return MAXALIGN(add_size(sizeof(int), 
+					mul_size(max_worker_processes,
+								sizeof(ContQueryProcGroup))));
+}
+
+void
+AdhocShmemInit(void)
+{
+	bool found = false;
+	AdhocShmem = ShmemInitStruct("AdhocShmem", adhoc_size_required(), &found);
+
+	if (found)
+		return;
+
+	MemSet(AdhocShmem, 0, adhoc_size_required());
+}
+
+/*
+ * Find a slot in the AdhocShmemStruct and initialize it if found.
+ * NULL is returned if there are no free slots.
+ */
+ContQueryProc *
+AdhocMgrGetProc()
+{
+	ContQueryProcGroup *grp = 0;
+	ContQueryProc *proc = 0;
+
+	int i = 0;
+	int ind = 0;
+
+	LWLockAcquire(AdhocMgrLock, LW_EXCLUSIVE);
+
+	for (i = 0; i < max_worker_processes; ++i)
+	{
+		ind = (AdhocShmem->counter + i) % max_worker_processes;
+
+		if (AdhocShmem->groups[ind].db_oid == 0)
+		{
+			grp = &AdhocShmem->groups[ind];
+			AdhocShmem->counter++;
+
+			break;
+		}
+	}
+
+	LWLockRelease(AdhocMgrLock);
+
+	if (grp)
+	{
+		MemSet(grp, 0, sizeof(ContQueryProcGroup));
+		grp->db_oid = MyDatabaseId;
+		grp->active = true;
+
+		proc = grp->procs;
+
+		proc->type = Scheduler;
```

Comment:
Not sure what's going on here. Why is this proc marked as a `Scheduler` proc? We should probably add a new enum for `Adhoc`?